extensions/qute/qute-reference-guide.adoc

```diff
@@ -0,0 +1,673 @@
+:toc:
+:toc-placement!:
+
+= Qute Reference Guide
+
+toc::[]
+
+== Hello World Example
+
+In this example, we'd like to demonstrate the basic workflow when working with Qute templates.
+Let's start with a simple hello world example.
+We will always need some *template contents*:
+
+.hello.html
+[source,html]
+----
+<html>
+ <p>Hello {name}! <1>
+</html>
+----
+<1> `{name}` is a value expression that is evaluated when the template is rendered.
+
+Then, we will need to parse the contents into a *template definition* Java object.
+A template definition is an instance of `io.quarkus.qute.Template`.
+
+If using Qute "standalone" you'll need to create an instance of `io.quarkus.qute.Engine` first.
+The `Engine` represents a central point for template management with dedicated configuration. 
+Let's use the convenient builder:
+
+[source,java]
+----
+Engine engine = Engine.builder().addDefaults().build();
+----
+
+TIP: In Quarkus, there is a preconfigured `Engine` available for injection - see <<quarkus_integration>>.
+
+If we have an `Engine` instance we could parse the template contents:
+
+[source,java]
+----
+Template helloTemplate = engine.parse(helloHtmlContent);
+----
+
+TIP: In Quarkus, you can simply inject the template definition. The template is automatically parsed and cached - see <<quarkus_integration>>.
+
+Finally, we will create a *template instance*, set the data and render the output:
+
+[source,java]
+----
+// Renders <html><p>Hello Jim!</p></html>
+helloTemplate.data("name", "Jim").render(); <1> 
+----
+<1> `Template.data(String, Object)` is a convenient method that creates a template instance and sets the data in one step.
+
+So the workflow is simple: 
+
+1. Create template contents (`hello.html`),
+2. Parse template definition (`io.quarkus.qute.Template`),
+3. Create template instance (`io.quarkus.qute.TemplateInstance`),
+4. Render output.
+
+The `Engine` is able to cache the definitions so that it's not necessary to parse the contents again and again.
+
+NOTE: In Quarkus, the caching is done automatically.
+
+== Core Features
+
+=== Syntax and Building Blocks
+
+The dynamic parts of a template include:
+
+* *comment* 
+** `{! This is a comment !}`,
+** could be multi-line,
+** may contain expressions and sections: `{! {#if true} !}`.
+* *expression*
+** outputs the evaluated value,
+** simple properties: `{foo}`, `{item.name}`,
+** virtual methods: `{item.get(name)}`, `{name ?: 'John'}`,
```

Comment:
Do we support, `{name ? 'something' : 'else'}` ? I don't see it in any example?

Also do we have a `concat` operator?

Reply:
> Do we support, {name ? 'something' : 'else'}

No, we don't.

> Also do we have a concat operator?

Not out of the box. What's the use case?

Reply:
I remember seeing a lot of templates which uses concat and also some sort of pipe operator to deal with transformations.

I think the latter is probably covered with the methods you can add? This was especially used to format dates. It would be nice to have something documented for that.

As for the former, it could typically be used when using a ternary operator if we end up having one :).

Thinking out loud, one other thing that I saw in other templating languages is some constructs to easily deal with plurals. But I suppose the elvis operator for simple cases and a ternary operator for more complex cases could do?

Reply:
For the record I added a value resolver to support `{condition ? 'something' : 'else'}`. We need to decide whether to test `true` or `truthy` (from JS). Currently, the condition must be true.